Remove sync for randperm on small tensors.

zasdfgbnm opened the pull request:
For small tensors, it is known that GPU operates slower than CPU. However, offloading to CPU causes host <--> device sync. As a result, although offloading to CPU has better microbenchmarks, it often hurts instead of benefits the end-to-end performance, and it could be a blocker for CUDA graphs. After discussion with @mcarilli and @ptrblck, we think it might be good to just remove this piece of code and let it be slow.

Microbenchmarks:

```python
def run50_sync(f):
    for _ in range(50):
        f()
    torch.cuda.synchronize()

torch.cuda.synchronize()
%timeit run50_sync(lambda: torch.randperm(3, device='cuda'))
%timeit run50_sync(lambda: torch.randperm(30, device='cuda'))
%timeit run50_sync(lambda: torch.randperm(300, device='cuda'))
%timeit run50_sync(lambda: torch.randperm(3000, device='cuda'))
%timeit run50_sync(lambda: torch.randperm(30000, device='cuda'))
%timeit run50_sync(lambda: torch.randperm(300000, device='cuda'))
%timeit run50_sync(lambda: torch.randperm(3000000, device='cuda'))
%timeit run50_sync(lambda: torch.randperm(30000000, device='cuda'))
```

Before this PR:
```
5.79 ms ± 51.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
5.78 ms ± 92.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
6.17 ms ± 87.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
9.65 ms ± 69.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
17.6 ms ± 133 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
21 ms ± 120 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
104 ms ± 880 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
944 ms ± 3.49 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

After this PR:
```
7.22 ms ± 11.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
7.28 ms ± 9.03 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
7.25 ms ± 10.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
9.19 ms ± 5.83 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
9.76 ms ± 162 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
12.3 ms ± 11.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
69.3 ms ± 42.3 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
716 ms ± 1.01 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

zasdfgbnm committed:
Remove sync for randperm on small tensors.

xuhdev commented on aten/src/ATen/native/cuda/TensorFactories.cu:
```diff
@@ -86,12 +81,6 @@ Tensor& randperm_out_cuda(Tensor& result, int64_t n, c10::optional<Generator> ge
 
   result.resize_({n});
 
-  if (n < 30000) {  // For small inputs, we offload it to CPU instead.
-    auto result_cpu = at::empty({n}, result.options().device(kCPU));
-    randperm_out(result_cpu, n, generator);
-    return result.copy_(result_cpu);
-  }
```
The performance loss seems to be significant. Do you have a more concrete example in which these lines hurt?

zasdfgbnm replied:
MaskRCNN uses randperm, and these lines prevent CUDA graph application of MaskRCNN

ngimel commented:
If you do `randint` instead of `random_`, and then smartly set `end_bit`, that would help perf a lot.

zasdfgbnm committed:
save

zasdfgbnm committed:
save

zasdfgbnm committed:
save

zasdfgbnm committed:
save

zasdfgbnm committed:
save

zasdfgbnm committed:
save

zasdfgbnm committed:
save

zasdfgbnm committed:
save

zasdfgbnm commented on test/test_tensor_creation_ops.py:
```diff
@@ -3129,16 +3129,15 @@ def test_rand(self, device, dtype):
             torch.rand(size, size, out=res2)
             self.assertEqual(res1, res2)
 
-    @slowTest
```
This test is not slow

zasdfgbnm committed:
save

zasdfgbnm committed:
save

zasdfgbnm commented:
@ngimel I have modified this PR according to your suggestion, the improvement is big, see updated benchmark in https://github.com/pytorch/pytorch/pull/54113#issue-594313054

zasdfgbnm committed:
Update cuda_to_hip_mappings.py

ngimel commented on aten/src/ATen/native/cuda/TensorFactories.cu:
```diff
@@ -82,69 +77,76 @@ Tensor empty_strided_cuda(IntArrayRef size, IntArrayRef stride, c10::optional<Sc
+  if (n == 0) {
+    return result;
+  } else if (bits <= 8) {
+    auto keys = at::empty(result.sizes(), opt.dtype(kByte)).random_(generator);
```
can you leave a comment that sorting on the first `bits` bits is still expected to produce all permutations with equal probability, provided good rng (that's true, right?)

zasdfgbnm replied:
For infinitely mantissa random number keys, yes, but for finite mantissa, then no. For example, if your key is `(1, 0, 1)`, then your sorted value will be `(1, 0, 2)` but not `(1, 2, 0)`. So the probability of value 0 appearing in front of value 1 is always larger than the probability of value 1 appearing in front of value 0. So this algorithm has some intrinsic flaw.

ngimel replied:
Hm, there's no mantissa, we are generating integers, no? Just random bits, and there should be all subsequences with random probability

zasdfgbnm replied:
I will add comment tomorrow.

ngimel replied:
But yeah, stable sort makes the argument more complex and possibly incorrect :-(

ngimel replied:
So looks like to avoid this problem you need bits >> log2(n), to minimize probability of generating equal numbers that result in biased sorting. I wonder if this algorithm can be saved

zasdfgbnm replied:
Yes, in theory, you need `bits == inf`. But let me theoretically analyze the probability to decide how many bits we should actually use.

zasdfgbnm replied:
So, to sample `N` numbers from a size `M` pool, there are in total `M^N` cases. And within these cases, there are `C(M, N)` non-duplicating cases. So the probability of getting nondegenerate keys (the correct case) is
```
P = C(M,N)/(M^N)
```

zasdfgbnm replied:
**TLDR:**
![image](https://user-images.githubusercontent.com/1032377/111667808-51314f00-87d2-11eb-9d56-f3dae9fa0081.png)


Full version: https://github.com/zasdfgbnm/things/blob/master/2021Q1/randperm.pdf

zasdfgbnm replied:
So, roughly, if you need char n, then you need short key; if you need short n, then you need int key; if you need int n, then you need long key. If you need long n, then you are wrong, and this is not supported by cub.

ngimel replied:
Ok, I think we can accept slightly skewed randperm :-) but you need more than `+1` that you have a as a safety margin for `bits`.

mcarilli replied:
It wouldn't be too hard to write a syncfree postprocessing kernel that additionally shuffles any islands of values whose keys match. Each thread reads its key and the keys of its neighbors to tell if it's part of an island. For each island, the "island leader" sees a key match at index i+1 but not index i-1. The island leader then reads more indices to the right (or communicates with neighboring threads through smem or shuffling) to figure out how big the island is. Most likely, the island will be very small, just a few values. The island leader then draws that many RNG and uses them to additionally shuffle values within the island and write them out.

The biggest fixed cost of this kernel would be a read of all the keys and values. Writing the reshuffled islands wouldn't be a big cost (thanks to your logic, there's a 90% chance there won't be any duplicate-key islands at all).

ngimel commented on aten/src/ATen/native/cuda/TensorFactories.cu:
```diff
@@ -82,69 +77,76 @@ Tensor empty_strided_cuda(IntArrayRef size, IntArrayRef stride, c10::optional<Sc
+  auto opt = TensorOptions().device(result.device());
+
+  int bits = std::min(64,
+    static_cast<int>(std::ceil(std::log2(static_cast<double>(n)))) + 1);
```
do you need +1 if you already have `ceil`? E.g. if `n=4` you need to sort on 2 bits?

zasdfgbnm replied:
See https://github.com/pytorch/pytorch/pull/54113#discussion_r596535150, I am just adding one extra mantissa to reduce the chance of two keys being the same, which will make the result more random.

ngimel commented on aten/src/ATen/native/cuda/TensorFactories.cu:
```diff
@@ -82,69 +77,76 @@ Tensor empty_strided_cuda(IntArrayRef size, IntArrayRef stride, c10::optional<Sc
+        range.data_ptr<scalar_t>(), reinterpret_cast<scalar_t*>(shuffled_data),
+        n, 0, bits);
+    });
+  } else {
```
this should never happen, given that you've errored out on n>int_max already?

zasdfgbnm replied:
Ahh, yes!

ngimel reviewed (commented):
This is awesome! Almost no perf loss, and graphable.

zasdfgbnm committed:
save

zasdfgbnm committed:
Merge branch 'master' of github.com:pytorch/pytorch into cub-randperm-nosync

zasdfgbnm committed:
Merge branch 'cub-randperm-nosync' of github.com:pytorch/pytorch into cub-randperm-nosync

zasdfgbnm committed:
save

zasdfgbnm committed:
save

zasdfgbnm committed:
save

zasdfgbnm committed:
save

zasdfgbnm committed:
save

zasdfgbnm committed:
save

zasdfgbnm committed:
Update TensorFactories.cu

zasdfgbnm committed:
Update TensorFactories.cu

zasdfgbnm commented:
@ngimel I think this should be ready.

The number of bits is computed according to the equation I get. Benchmark result using this equation is updated in https://github.com/pytorch/pytorch/pull/54113#issue-594313054.

I also added comments. I tried to upload my document to arxiv, but they rejected it because it was not a research paper, so I switch to another preprint server.

The int64 branch is still needed, because, for large n, my equation shows that it needs 64bits.

I changed the test from `slowTest` to `onlyCUDA`, because it is not slow, but flaky on a CPU-only build.

zasdfgbnm committed:
save

zasdfgbnm committed:
save

zasdfgbnm committed:
save

zasdfgbnm committed:
save

zasdfgbnm committed:
save

zasdfgbnm committed:
cleanup

mcarilli commented on aten/src/ATen/native/cuda/TensorFactories.cuh:
```diff
@@ -0,0 +1,55 @@
+  for (int i = island_size - 1; i > 0; i--) {
+    unsigned int r = curand(&state) % (i + 1);
+    if (i != r) {
+      scalar_t tmp = data[i];
```
Shouldn't you be accessing based on tid here?

zasdfgbnm replied:
There is a
```C++
data += tid;
```
in https://github.com/pytorch/pytorch/pull/54113/files/bda1c838d1b85a54562ad85fa508a3a434f29589#diff-db9fbcdb96e426b3b546a6fe794fb864e74cebb1cf31ec70ccd97b7db364ad37R26

mcarilli commented on aten/src/ATen/native/cuda/TensorFactories.cuh:
```diff
@@ -0,0 +1,55 @@
+template<typename T, typename scalar_t>
+void randperm_handle_duplicate_keys(T *keys, scalar_t *data, int bits, int64_t n, c10::optional<at::Generator> gen_) {
+  auto gen = at::get_generator_or_default<at::CUDAGeneratorImpl>(gen_, at::cuda::detail::getDefaultCUDAGenerator());
+  int64_t counter_offset = n;
```
:joy:

mcarilli commented on aten/src/ATen/native/cuda/TensorFactories.cu:
```diff
@@ -79,72 +75,123 @@ Tensor empty_strided_cuda(IntArrayRef size, IntArrayRef stride, c10::optional<Sc
+// This thread considers itself the "island leader". The island leader then reads more indices to
+// the right to figure out how big the island is. Most likely, the island will be very small,
+// just a few values. The island leader then rolls that many RNG, uses them to additionally
+// shuffle values within the island, and writes them out.
```
`additionally shuffle values using Fisher-Yates`

zasdfgbnm replied:
fixed

zasdfgbnm committed:
Update TensorFactories.cu

ngimel commented:
Can you please rebase? All windows builds are broken in the base.

zasdfgbnm committed:
Merge branch 'master' of github.com:pytorch/pytorch into cub-randperm-nosync

zasdfgbnm committed:
Merge branch 'cub-randperm-nosync' of github.com:pytorch/pytorch into cub-randperm-nosync

zasdfgbnm commented:
@ngimel rebased

zasdfgbnm committed:
Merge branch 'master' of github.com:pytorch/pytorch into cub-randperm-nosync

zasdfgbnm commented:
Failing on MSVC:
```
Traceback (most recent call last):
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_device_type.py", line 292, in instantiated_test
    result = test_fn(self, *args)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_device_type.py", line 701, in only_fn
    return fn(slf, device, *args, **kwargs)
  File "test_tensor_creation_ops.py", line 3151, in test_randperm
    self.assertEqual(res1, res2, atol=0, rtol=0)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_utils.py", line 1251, in assertEqual
    super().assertTrue(result, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
AssertionError: False is not true : Tensors failed to compare as equal!With rtol=0 and atol=0, found 2 element(s) (out of 50000) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 33419.0 (3176.0 vs. 36595.0), which occurred at index 49999.
```
interesting

zasdfgbnm committed:
debug

zasdfgbnm committed:
Update TensorFactories.cuh

zasdfgbnm committed:
Update cuda_distributions_test.cu

zasdfgbnm commented:
Looks like the cleanup kernel is the cause of flaky test on MSVC, so I have to disable it on MSVC. Otherwise, this PR is good to go.

zasdfgbnm committed:
Merge branch 'master' into cub-randperm-nosync

zasdfgbnm committed:
fix

zasdfgbnm committed:
save

zasdfgbnm committed:
Update cuda_distributions_test.cu

zasdfgbnm committed:
Merge branches 'master' and 'master' of github.com:pytorch/pytorch into cub-randperm-nosync

zasdfgbnm committed:
stream

zasdfgbnm committed:
no-comflict

zasdfgbnm committed:
Merge branch 'master' of github.com:pytorch/pytorch into cub-randperm-nosync

ngimel commented:
This one needs to be rebased now.

zasdfgbnm committed:
Merge branch 'master' of github.com:pytorch/pytorch into cub-randperm-nosync

zasdfgbnm commented:
This is rebased, but I don't know if tests are running or not: https://github.com/pytorch/pytorch/issues/55683, I can not locally run tests.

zasdfgbnm committed:
Merge branch 'master' of github.com:pytorch/pytorch into cub-randperm-nosync

zasdfgbnm commented:
rebased again, tests should be running

zasdfgbnm committed:
Update Randperm.cu

zasdfgbnm committed:
try

zasdfgbnm committed:
save

zasdfgbnm committed:
fix

zasdfgbnm committed:
debug

zasdfgbnm committed:
Merge branch 'master' of github.com:pytorch/pytorch into cub-randperm-nosync

zasdfgbnm committed:
more

zasdfgbnm committed:
revert

zasdfgbnm committed:
save

zasdfgbnm committed:
save

zasdfgbnm commented:
@ngimel @mcarilli `test_randperm_cuda` on Windows is still failing. I don't know the reason for the failing, but here is what I observe:
1. It fails on all our Windows CIs (10.1, 11.1, 11.2), but I can not reproduce it locally (also 11.2 on Windows)
2. The failure is caused by the kernel `randperm_handle_duplicate_keys_kernel`. This kernel is causing `out1 = randperm(n)` and `randperm(n, out=out2)` generating different values.
3. By adding prints, I can see strange behaviors that, although nothing in the kernel prints `tid<=20`, the 19th element of the `values` tensor is still permuted with the 20th element.
4. The test is still failing, even if I change the `unsigned int r = curand(&state) % (i + 1);` into `unsigned int r = 0`, and remove the three lines that initialize philox state(but still keeping the `at::PhiloxCudaState philox_args` argument)
5. If I remove the `at::PhiloxCudaState philox_args` argument, the test stops failing. But if I keep this argument, even if I am not using this argument in the kernel, the test will fail. See https://github.com/pytorch/pytorch/pull/56344/files#diff-3439e602b04b41a8cbe64924037b47250b2e026d56bf16aa2948450e5e18fa85 vs https://github.com/pytorch/pytorch/pull/56349/files#diff-3439e602b04b41a8cbe64924037b47250b2e026d56bf16aa2948450e5e18fa85

Based on the above analysis, I think this is a compiler bug. I guess some optimization pass invoked for maxwell transforms the code to some complete nonsense. For the next step, my suggestions are:
1. You two, please review the `randperm_handle_duplicate_keys_kernel` to see if you can find any mistakes.
2. If no mistake, I will disable this kernel on Windows, and let this PR go.
3. I will open an issue and maybe continue debugging this after this PR is merged.

zasdfgbnm commented:
One question, to check my knowledge: If I write `data[0]` then immediately read `data[0]` in the same thread, I don't need a thread fence. Am I correct?

ngimel commented:
> One question, to check my knowledge: If I write `data[0]` then immediately read `data[0]` in the same thread, I don't need a thread fence. Am I correct?

Yes, that's correct.

zasdfgbnm committed:
skip msvc

zasdfgbnm committed:
Merge branch 'master' of github.com:pytorch/pytorch into cub-randperm-nosync

zasdfgbnm committed:
remove debug prints

zasdfgbnm committed:
save

zasdfgbnm committed:
save

zasdfgbnm committed:
save

ngimel commented:
@zasdfgbnm can you please rebase, there's a conflict?

zasdfgbnm committed:
Merge branch 'master' into cub-randperm-nosync

zasdfgbnm commented:
@ngimel rebased
